rfc/003-ed25519-verification.md

```diff
@@ -0,0 +1,41 @@
+# RFC 003: Ed25519 Verification
+
+## Changelog
+
+- August 21, 2020: initialized
+
+## Author(s)
+
+- Marko (@marbar3778)
+
+## Context
+
+Ed25519 keys are the only supported key types for Tendermint validators currently. Tendermint-Go wraps the ed25519 key implementation from the go standard library. As more clients are implemented to communicate with the canonical Tendermint implementation (Tendermint-Go) different implementations of ed25519 will be used. Due to [RFC 8032](https://www.rfc-editor.org/rfc/rfc8032.html) not guaranteeing implementation compatibility, Tendermint clients must to come to an agreement of how to guarantee implementation compatibility. [Zcash](https://z.cash/) has multiple implementations of their client and have identified this as a problem as well. The team at Zcash has made a proposal to address this issue, [Zcash improvement proposal 215](https://zips.z.cash/zip-0215).
+
+## Proposal
+
+- Tendermint-Go would adopt [hdevalence/ed25519consensus](https://github.com/hdevalence/ed25519consensus).
+  - This library is implemented as an extension of the go standard library one.
+- Tendermint-rs would adopt [ed25519-zebra](https://github.com/ZcashFoundation/ed25519-zebra)
+  - related [issue](https://github.com/informalsystems/tendermint-rs/issues/355)
+
+As signature verification is one of the major bottlenecks of Tendermint-go, if ZIP 215 is adopted batch verification of signatures will be safe in consensus critical areas.
+
+## Status
+
+Proposed
+
+## Consequences
+
+### Positive
+
+- Batch verification
```

Comment:
hdevalence/ed25519consensus doesn't support batch verification: [godoc](https://pkg.go.dev/github.com/hdevalence/ed25519consensus)
This library is a good example with ed25519 batch verification: https://github.com/oasisprotocol/ed25519

Reply:
Will be more explicit, this will come at a later date. 

As for oasis protocol library, we discussed and would only use for our fast sync process and not for consensus. The verification function in `hdevalence/ed25519consensus` handles sig verification in more detail, thus we would be able to utilize it for consensus verification. Since this is not supported at this time I decided to leave it out of this conversation. Batch verification is an implementation level decision as well, no need to have the conversation in the spec repo.

Reply:
So let's have a section  _further ideas_ and let's put it there.

Reply:
FWIW, I think it makes sense to classify the *ability* to perform batch verification as a "now" positive rather than in a "further ideas" section, because it's something that's enabled immediately by the proposed change, whereas "further ideas" seems like a place for changes that are not proposed now but could be made in the future.   In the case of batch verification, if ZIP215 were adopted, using batch verification in Tendermint-Go wouldn't require any further changes to the spec, it would be purely an implementation optimization.

(Of course, this is just my opinion, and I don't have full context on Tendermint or its processes, so please feel free to discount it appropriately)